File: src/player/mediaperch/protocol.cpp
// SPDX-License-Identifier: GPL-3.0-or-later

#include "mediaperch/protocol.hpp"

#include <cstring>

namespace mp::ipc {
namespace {

/// The largest string this will build. A path is a few hundred bytes; a
/// megabyte is a length somebody made up.
constexpr std::uint32_t k_max_string = k_max_payload;

/// How many items a count field may claim before it is a claim rather than a
/// count. Every list here is a playlist, a settings tree or a log tail.
constexpr std::uint32_t k_max_items = 1u << 16;

} // namespace

const char* state_name(State s) noexcept
{
    switch (s) {
    case State::stopped:
        return "stopped";
    case State::playing:
        return "playing";
    case State::paused:
        return "paused";
    }
    return "?";
}

const char* kind_name(Kind k) noexcept
{
    switch (k) {
    case Kind::hello:
        return "hello";
    case Kind::status:
        return "status";
    case Kind::play:
        return "play";
    case Kind::enqueue:
        return "enqueue";
    case Kind::clear:
        return "clear";
    case Kind::pause:
        return "pause";
    case Kind::resume:
        return "resume";
    case Kind::stop:
        return "stop";
    case Kind::seek:
        return "seek";
    case Kind::next:
        return "next";
    case Kind::previous:
        return "previous";
    case Kind::playlist:
        return "playlist";
    case Kind::settings:
        return "settings";
    case Kind::setting_set:
        return "setting_set";
    case Kind::log:
        return "log";
    case Kind::subscribe:
        return "subscribe";
    case Kind::quit:
        return "quit";
    case Kind::save:
        return "save";
    case Kind::ok:
        return "ok";
    case Kind::error:
        return "error";
    case Kind::hello_reply:
        return "hello_reply";
    case Kind::status_reply:
        return "status_reply";
    case Kind::playlist_reply:
        return "playlist_reply";
    case Kind::settings_reply:
        return "settings_reply";
    case Kind::log_reply:
        return "log_reply";
    case Kind::event_state:
        return "event_state";
    case Kind::event_log:
        return "event_log";
    }
    return "unknown";
}

// --------------------------------------------------------------------------

void Writer::u8(std::uint8_t v)
{
    out_.push_back(v);
}

void Writer::u32(std::uint32_t v)
{
    for (int i = 0; i < 4; ++i) {
        out_.push_back(static_cast<std::uint8_t>((v >> (8 * i)) & 0xFFu));
    }
}

void Writer::u64(std::uint64_t v)
{
    for (int i = 0; i < 8; ++i) {
        out_.push_back(static_cast<std::uint8_t>((v >> (8 * i)) & 0xFFu));
    }
}

void Writer::i64(std::int64_t v)
{
    u64(static_cast<std::uint64_t>(v));
}

void Writer::f64(double v)
{
    // Through the bits rather than through text: a double that survives a round
    // trip as decimal is a double somebody rounded.
    std::uint64_t bits = 0;
    std::memcpy(&bits, &v, sizeof(bits));
    u64(bits);
}

void Writer::str(std::string_view v)
{
    const auto n = static_cast<std::uint32_t>(v.size());
    u32(n);
    out_.insert(out_.end(), v.begin(), v.end());
}

void Writer::format(const Format& f)
{
    u32(f.sample_rate);
    u32(f.channels);
    u32(f.channel_mask);
    u32(static_cast<std::uint32_t>(f.sample_type));
    u32(static_cast<std::uint32_t>(f.encoding));
    u32(f.valid_bits);
}

// --------------------------------------------------------------------------

bool Reader::take(std::size_t n) noexcept
{
    if (bad_ || size_ - at_ < n) {
        bad_ = true;
        return false;
    }
    return true;
}

std::uint8_t Reader::u8() noexcept
{
    if (!take(1)) {
        return 0;
    }
    return data_[at_++];
}

std::uint32_t Reader::u32() noexcept
{
    if (!take(4)) {
        return 0;
    }
    std::uint32_t v = 0;
    for (int i = 0; i < 4; ++i) {
        v |= static_cast<std::uint32_t>(data_[at_ + static_cast<std::size_t>(i)]) << (8 * i);
    }
    at_ += 4;
    return v;
}

std::uint64_t Reader::u64() noexcept
{
    if (!take(8)) {
        return 0;
    }
    std::uint64_t v = 0;
    for (int i = 0; i < 8; ++i) {
        v |= static_cast<std::uint64_t>(data_[at_ + static_cast<std::size_t>(i)]) << (8 * i);
    }
    at_ += 8;
    return v;
}

std::int64_t Reader::i64() noexcept
{
    return static_cast<std::int64_t>(u64());
}

double Reader::f64() noexcept
{
    const std::uint64_t bits = u64();
    double v = 0.0;
    std::memcpy(&v, &bits, sizeof(v));
    return v;
}

std::string Reader::str()
{
    const std::uint32_t n = u32();
    // Checked before it is trusted with an allocation. A length is the one
    // field in a message that can cost something on its own.
    if (n > k_max_string || !take(n)) {
        bad_ = true;
        return {};
    }
    std::string out(reinterpret_cast<const char*>(data_ + at_), n);
    at_ += n;
    return out;
}

Format Reader::format() noexcept
{
    Format f{};
    f.sample_rate = u32();
    f.channels = u32();
    f.channel_mask = u32();
    f.sample_type = static_cast<SampleType>(u32());
    f.encoding = static_cast<Encoding>(u32());
    f.valid_bits = u32();
    return f;
}

// --------------------------------------------------------------------------

std::vector<std::uint8_t> frame(Kind kind, std::uint32_t id, const Writer& payload)
{
    Writer head;
    head.u32(k_magic);
    // Version and kind are 16 bits each, written in the same order the header
    // declares them.
    head.u8(static_cast<std::uint8_t>(k_version & 0xFFu));
    head.u8(static_cast<std::uint8_t>((k_version >> 8) & 0xFFu));
    const auto k = static_cast<std::uint16_t>(kind);
    head.u8(static_cast<std::uint8_t>(k & 0xFFu));
    head.u8(static_cast<std::uint8_t>((k >> 8) & 0xFFu));
    head.u32(id);
    head.u32(static_cast<std::uint32_t>(payload.size()));

    std::vector<std::uint8_t> out = head.bytes();
    out.insert(out.end(), payload.bytes().begin(), payload.bytes().end());
    return out;
}

std::vector<std::uint8_t> frame(Kind kind, std::uint32_t id)
{
    const Writer empty;
    return frame(kind, id, empty);
}

bool parse_header(const std::uint8_t* data, std::size_t size, Header& out) noexcept
{
    if (data == nullptr || size < k_header_bytes) {
        return false;
    }
    Reader r{data, k_header_bytes};
    out.magic = r.u32();
    // One call per statement: the order of two reads either side of an operator
    // is not something C++ promises, and a header parsed backwards on one
    // compiler is exactly the kind of bug this format exists to avoid.
    const std::uint8_t version_low = r.u8();
    const std::uint8_t version_high = r.u8();
    const std::uint8_t kind_low = r.u8();
    const std::uint8_t kind_high = r.u8();
    out.version = static_cast<std::uint16_t>(version_low | (version_high << 8));
    out.kind = static_cast<std::uint16_t>(kind_low | (kind_high << 8));
    out.id = r.u32();
    out.payload = r.u32();
    if (!r.ok() || out.magic != k_magic) {
        return false;
    }
    // A version mismatch is not a message this build can read *at all*: the
    // fields may have moved. Saying no here is better than a shell and an
    // engine agreeing on a length and disagreeing on everything else.
    if (out.version != k_version || out.payload > k_max_payload) {
        return false;
    }
    return true;
}

void write(Writer& w, const Status& s)
{
    w.u32(static_cast<std::uint32_t>(s.state));
    w.u32(s.index);
    w.u32(s.count);
    w.u64(s.position);
    w.u64(s.length);
    w.str(s.track);
    w.str(s.decoder);
    w.str(s.device);
    w.format(s.source);
    w.format(s.wire);
    w.u32(s.fidelity);
    w.u8(s.processed ? 1u : 0u);
    w.u64(s.frames_rendered);
    w.u64(s.underruns);
    w.str(s.error);
}

bool read(Reader& r, Status& s)
{
    s.state = static_cast<State>(r.u32());
    s.index = r.u32();
    s.count = r.u32();
    s.position = r.u64();
    s.length = r.u64();
    s.track = r.str();
    s.decoder = r.str();
    s.device = r.str();
    s.source = r.format();
    s.wire = r.format();
    s.fidelity = r.u32();
    s.processed = r.u8() != 0;
    s.frames_rendered = r.u64();
    s.underruns = r.u64();
    s.error = r.str();
    return r.ok();
}

void write(Writer& w, const std::vector<Setting>& settings)
{
    w.u32(static_cast<std::uint32_t>(settings.size()));
    for (const Setting& s : settings) {
        w.str(s.key);
        w.str(s.value);
        w.str(s.description);
    }
}

bool read(Reader& r, std::vector<Setting>& settings)
{
    const std::uint32_t n = r.u32();
    if (!r.ok() || n > k_max_items) {
        return false;
    }
    settings.clear();
    settings.reserve(n);
    for (std::uint32_t i = 0; i < n; ++i) {
        Setting s;
        s.key = r.str();
        s.value = r.str();
        s.description = r.str();
        if (!r.ok()) {
            return false;
        }
        settings.push_back(std::move(s));
    }
    return r.ok();
}

void write_strings(Writer& w, const std::vector<std::string>& items)
{
    w.u32(static_cast<std::uint32_t>(items.size()));
    for (const std::string& s : items) {
        w.str(s);
    }
}

bool read_strings(Reader& r, std::vector<std::string>& items)
{
    const std::uint32_t n = r.u32();
    if (!r.ok() || n > k_max_items) {
        return false;
    }
    items.clear();
    items.reserve(n);
    for (std::uint32_t i = 0; i < n; ++i) {
        items.push_back(r.str());
        if (!r.ok()) {
            return false;
        }
    }
    return r.ok();
}

} // namespace mp::ipc

File: shell/winui/Ipc/Protocol.cs
// SPDX-License-Identifier: GPL-3.0-or-later

namespace MediaPerch.Shell.Ipc;

/// <summary>
/// The other half of <c>src/player/mediaperch/protocol.hpp</c>.
/// </summary>
/// <remarks>
/// <b>Two descriptions of one wire, and that is a cost this takes on
/// deliberately.</b> The alternative was a serialiser or a generator, and §10's
/// framing is small enough that either would be more machinery than the thing
/// it describes. What keeps them in step is that the format is versioned and
/// the engine answers <c>error</c> to a kind it does not know -- so a shell that
/// is ahead of its engine is told, rather than reading a field that moved.
/// </remarks>
public static class Protocol
{
    public const uint Magic = 0x5049504Du;
    // 2: a settings row carries its kind, choices, group, `when` and hints.
    public const ushort Version = 2;
    public const uint MaxPayload = 1u << 20;
    public const int HeaderBytes = 16;

    /// <summary>The engine's default pipe, matching <c>default_pipe_name</c>.</summary>
    public const string DefaultPipe = @"mediaperch";
}

/// <summary>
/// What a message is. Numbered in three ranges so that reading a trace tells
/// you which direction it was going without a table.
/// </summary>
public enum Kind : ushort
{
    // requests, shell to engine
    Hello = 1,
    Status = 2,
    Play = 3,
    Enqueue = 4,
    Clear = 5,
    Pause = 6,
    Resume = 7,
    Stop = 8,
    Seek = 9,
    Next = 10,
    Previous = 11,
    Playlist = 12,
    Settings = 13,
    SettingSet = 14,
    Log = 15,
    Subscribe = 16,
    // **Mirrored from protocol.hpp, and held to it by a test** -- the C++ enum
    // is the one description and tests/protocol_test.cpp reads this file
    // against it. These two were once the other way round here, and "Save
    // settings" quit the engine while closing the window asked it to save.
    Quit = 17,
    Save = 18,
    Calibrate = 19,
    Profile = 20,
    Display = 21,
    Graph = 22,
    NodeSettings = 23,
    NodeSettingSet = 24,
    Modules = 25,
    EngineSettings = 26,
    EngineSettingSet = 27,
    Surface = 28,
    PlayAt = 29,
    MoveEntry = 30,

    // replies, engine to shell
    Ok = 128,
    Error = 129,
    HelloReply = 130,
    StatusReply = 131,
    PlaylistReply = 132,
    SettingsReply = 133,
    LogReply = 134,
    ProfileReply = 135,
    GraphReply = 136,
    NodeSettingsReply = 137,
    ModulesReply = 138,
    EngineSettingsReply = 139,
    SurfaceReply = 140,

    // events, engine to shell, unasked
    EventState = 200,
    EventLog = 201,
}

public enum PlayState : uint
{
    Stopped = 0,
    Playing = 1,
    Paused = 2,
}

/// <summary>What a node is. Numbers rather than words, as the wire has them.</summary>
public enum NodeKind : uint
{
    Source = 0,
    Convert = 1,
    Dsp = 2,
    Sink = 3,
    VideoSource = 4,
    VideoStage = 5,
    Presenter = 6,
}

[Flags]
public enum NodeFlags : uint
{
    None = 0,
    Removable = 1 << 0,
    Settable = 1 << 1,
}

public readonly record struct Header(uint Magic, ushort Version, Kind Kind, uint Id,
                                       uint Payload);

public sealed record Node(string Id, NodeKind Kind, string Module, string Name,
                            NodeFlags Flags);

public sealed record Edge(string From, string To);

public sealed class Graph
{
    public List<Node> Nodes { get; } = new();
    public List<Edge> Edges { get; } = new();
}

/// <summary>
/// What kind of value a settings row takes -- <c>SettingKind</c> in
/// protocol.hpp, and held to it by the same test that holds <c>Kind</c>.
/// </summary>
/// <remarks>
/// <b>Advice about drawing, not a rule about values.</b> The engine validates
/// and this shell does not: a choice is an editable drop-down, a number field
/// has no minimum and no maximum, and whatever was typed is sent. What the
/// kind decides is which control to draw, which is the one decision a shell
/// cannot make from the text.
/// </remarks>
public enum SettingKind : uint
{
    Text = 0,
    Choice = 1,
    Integer = 2,
    Number = 3,
    Toggle = 4,
    Size = 5,
    Path = 6,
}

/// <summary>One row of a module's own settings.</summary>
/// <param name="ReadOnly">
/// A measurement rather than a setting -- a peak, a cost, a latency, a handle.
/// The engine decides this, from the <c>(read only)</c> every module's
/// <c>describe</c> ends such a row with; a shell that decided by looking for
/// those two words would be parsing English over a wire.
/// </param>
/// <param name="Kind">What the value is, so a control can be chosen.</param>
/// <param name="Choices">The words a <see cref="SettingKind.Choice"/> offers.</param>
/// <param name="Group">The heading the row sits under; empty is the top.</param>
/// <param name="When">
/// <c>key=value[,value...]</c>: the row matters only while the named row has
/// one of those values, and is folded away otherwise. Empty is always.
/// </param>
/// <param name="Hints">
/// <c>word=value</c> pairs separated by spaces -- <c>min=</c>, <c>max=</c>,
/// <c>step=</c>, <c>unit=</c>, <c>pick=folder</c> -- as the module wrote them.
/// </param>
public sealed record Setting(string Key, string Value, string Description,
                               bool ReadOnly, SettingKind Kind, List<string> Choices,
                               string Group, string When, string Hints)
{
    /// <summary>One hint's value, or null when the module did not give it.</summary>
    public string? Hint(string name)
    {
        foreach (string word in Hints.Split(' ', StringSplitOptions.RemoveEmptyEntries))
        {
            int equals = word.IndexOf('=');
            if (equals > 0 && word.AsSpan(0, equals).SequenceEqual(name))
            {
                return word[(equals + 1)..];
            }
        }
        return null;
    }
}

public sealed record ModuleRow(uint Kind, string Id, string Name, uint Priority,
                                 bool Allowed);

public sealed record Format(uint SampleRate, uint Channels, uint ChannelMask,
                              uint SampleType, uint Encoding, uint ValidBits);

public sealed class Status
{
    public PlayState State { get; init; }
    public uint Index { get; init; }
    public uint Count { get; init; }
    /// <summary>
    /// What <see cref="Position"/>, <see cref="Length"/> and
    /// <see cref="ItemPosition"/> count in: the source's rate while audio
    /// plays, and the video engine's own clock's (milliseconds) while a
    /// picture with no audio plays alone.
    /// </summary>
    public uint ClockRate { get; init; }
    public ulong Position { get; init; }
    public ulong Length { get; init; }
    /// <summary>
    /// How far into this track, where <see cref="Position"/> is how far into
    /// the queue. The two are in different units: a gapless queue is one stream
    /// to the device, so the position counts straight through boundaries.
    /// </summary>
    public ulong ItemPosition { get; init; }
    public string Track { get; init; } = string.Empty;
    public string Decoder { get; init; } = string.Empty;
    public string Device { get; init; } = string.Empty;
    public Format Source { get; init; } = new(0, 0, 0, 0, 0, 0);
    public Format Wire { get; init; } = new(0, 0, 0, 0, 0, 0);
    public uint Fidelity { get; init; }
    public bool Processed { get; init; }
    public ulong FramesRendered { get; init; }
    public ulong Underruns { get; init; }
    public string Error { get; init; } = string.Empty;
}

/// <summary>Decoding, kept beside the types it decodes.</summary>
public static class Decode
{
    public static Format ReadFormat(Reader r) =>
        new(r.U32(), r.U32(), r.U32(), r.U32(), r.U32(), r.U32());

    public static Status ReadStatus(Reader r)
    {
        var state = (PlayState)r.U32();
        uint index = r.U32();
        uint count = r.U32();
        uint clockRate = r.U32();
        ulong position = r.U64();
        ulong length = r.U64();
        ulong itemPosition = r.U64();
        string track = r.Str();
        string decoder = r.Str();
        string device = r.Str();
        Format source = ReadFormat(r);
        Format wire = ReadFormat(r);
        uint fidelity = r.U32();
        bool processed = r.U8() != 0;
        ulong rendered = r.U64();
        ulong underruns = r.U64();
        string error = r.Str();
        return new Status
        {
            State = state,
            Index = index,
            Count = count,
            ClockRate = clockRate,
            Position = position,
            Length = length,
            ItemPosition = itemPosition,
            Track = track,
            Decoder = decoder,
            Device = device,
            Source = source,
            Wire = wire,
            Fidelity = fidelity,
            Processed = processed,
            FramesRendered = rendered,
            Underruns = underruns,
            Error = error,
        };
    }

    public static List<Setting> ReadSettings(Reader r)
    {
        var out_ = new List<Setting>();
        uint n = r.U32();
        if (!r.Ok || n > 1u << 16)
        {
            return out_;
        }
        for (uint i = 0; i < n; ++i)
        {
            string key = r.Str();
            string value = r.Str();
            string description = r.Str();
            bool readOnly = r.U8() != 0;
            uint kind = r.U32();
            List<string> choices = ReadStrings(r);
            string group = r.Str();
            string when = r.Str();
            string hints = r.Str();
            // A kind from a newer engine is a box, which is what every row was
            // drawn as before there were kinds.
            out_.Add(new Setting(key, value, description, readOnly,
                                 kind <= (uint)SettingKind.Path ? (SettingKind)kind
                                                                : SettingKind.Text,
                                 choices, group, when, hints));
        }
        return out_;
    }

    public static List<string> ReadStrings(Reader r)
    {
        var out_ = new List<string>();
        uint n = r.U32();
        if (!r.Ok || n > 1u << 16)
        {
            return out_;
        }
        for (uint i = 0; i < n; ++i)
        {
            out_.Add(r.Str());
        }
        return out_;
    }

    public static Graph ReadGraph(Reader r)
    {
        var graph = new Graph();
        uint nodes = r.U32();
        if (!r.Ok || nodes > 1u << 16)
        {
            return graph;
        }
        for (uint i = 0; i < nodes; ++i)
        {
            string id = r.Str();
            var kind = (NodeKind)r.U32();
            string module = r.Str();
            string name = r.Str();
            var flags = (NodeFlags)r.U32();
            graph.Nodes.Add(new Node(id, kind, module, name, flags));
        }
        uint edges = r.U32();
        if (!r.Ok || edges > 1u << 16)
        {
            return graph;
        }
        for (uint i = 0; i < edges; ++i)
        {
            graph.Edges.Add(new Edge(r.Str(), r.Str()));
        }
        return graph;
    }

    public static List<ModuleRow> ReadModules(Reader r)
    {
        var out_ = new List<ModuleRow>();
        uint n = r.U32();
        if (!r.Ok || n > 1u << 16)
        {
            return out_;
        }
        for (uint i = 0; i < n; ++i)
        {
            out_.Add(new ModuleRow(r.U32(), r.Str(), r.Str(), r.U32(), r.U8() != 0));
        }
        return out_;
    }
}
